sdk/python/feast/infra/utils/aws_utils.py

```diff
@@ -0,0 +1,286 @@
+import os
+import tempfile
+import uuid
+from typing import Tuple
+
+import pandas as pd
+import pyarrow as pa
+import pyarrow.parquet as pq
+from tenacity import retry, retry_if_exception_type, wait_exponential
+
+from feast.errors import RedshiftCredentialsError, RedshiftQueryError
+from feast.type_map import pa_to_redshift_value_type
+
+try:
+    import boto3
+    from botocore.config import Config
+    from botocore.exceptions import ClientError
+except ImportError as e:
+    from feast.errors import FeastExtrasDependencyImportError
+
+    raise FeastExtrasDependencyImportError("aws", str(e))
+
+
+def get_redshift_data_client(aws_region: str):
+    """
+    Get the Redshift Data API Service client for the given AWS region.
+    """
+    return boto3.client("redshift-data", config=Config(region_name=aws_region))
+
+
+def get_s3_resource(aws_region: str):
+    """
+    Get the S3 resource for the given AWS region.
+    """
+    return boto3.resource("s3", config=Config(region_name=aws_region))
+
+
+def get_bucket_and_key(s3_path: str) -> Tuple[str, str]:
+    """
+    Get the S3 bucket and key given the full path.
+
+    For example get_bucket_and_key("s3://foo/bar/test.file") returns ("foo", "bar/test.file")
+
+    If the s3_path doesn't start with "s3://", it throws ValueError.
+    """
+    assert s3_path.startswith("s3://")
```

Comment:
Can we please raise an exception instead of using `assert`? I am not sure if it's a good practice to use `assert` outside of tests since it can be disabled when optimizations are turned on.

Reply:
I didn't realize that! I'll change this, although we have couple of other places where we use assert too.

Reply:
I know we have type assertions for Mypy. Do we have others as well?

Reply:
Ah you're right mostly it's type assertions. There are couple of them which do actual checks though:
```
$ pygrep -r "assert " sdk/python/feast | grep -v isinstance | grep -v type
sdk/python/feast/infra/online_stores/sqlite.py:        assert (
sdk/python/feast/infra/local.py:        assert config is not None
sdk/python/feast/infra/utils/aws_utils.py:    assert s3_path.startswith("s3://")
sdk/python/feast/staging/storage_client.py:        assert remote_path_prefix is not None and remote_path_suffix is not None
sdk/python/feast/staging/storage_client.py:        assert remote_path.hostname is not None
sdk/python/feast/staging/storage_client.py:        assert remote_path.hostname is not None
sdk/python/feast/staging/storage_client.py:        assert uri.hostname == f"{self.account_name}.blob.core.windows.net"
sdk/python/feast/staging/storage_client.py:        assert uri.username
sdk/python/feast/feature_view.py:        assert _input is not None
sdk/python/feast/repo_config.py:        assert "provider" in values
sdk/python/feast/repo_config.py:        assert "provider" in values
```

Reply:
Ok, well we do need to clean that up as well. But I think in some cases we have Optional[foo] being asserted into `foo` by assert statements, so it may also be for MyPy

Reply:
Yeah that's true. I'll open a separate task for that.